misk/src/test/kotlin/misk/hibernate/ManualCharacterQuery.kt

```diff
@@ -0,0 +1,165 @@
+package misk.hibernate
+
+import misk.hibernate.TransacterTest.ActorAndReleaseDate
+import misk.hibernate.TransacterTest.CharacterQuery
+import misk.hibernate.TransacterTest.MovieNameAndReleaseDate
+import java.time.LocalDate
+import javax.inject.Singleton
+import javax.persistence.criteria.CompoundSelection
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.Path
+import javax.persistence.criteria.Predicate
+import javax.persistence.criteria.Root
+import kotlin.reflect.KClass
+
+// TODO(jwilson): build this with metaprogramming (reflection/codegen) instead of code.
+class ManualCharacterQuery : CharacterQuery {
+  private val constraints = mutableListOf<ConstraintSpec>()
+
+  private fun addConstraint(path: List<String>, predicate: PredicateSpec): CharacterQuery {
+    constraints.add(ConstraintSpec(path, predicate))
+    return this
+  }
+
+  override fun name(name: String) = addConstraint(
+      path = listOf("name"),
+      predicate = PredicateSpec.Eq(name))
+
+  override fun actorName(name: String) = addConstraint(
+      path = listOf("actor", "name"),
+      predicate = PredicateSpec.Eq(name))
+
+  override fun movieName(name: String) = addConstraint(
+      path = listOf("movie", "name"),
+      predicate = PredicateSpec.Eq(name))
+
+  override fun movieReleaseDateBefore(upperBound: LocalDate) = addConstraint(
+      path = listOf("movie", "release_date"),
+      predicate = PredicateSpec.Lt(upperBound))
+
+  override fun uniqueResult(session: Session): DbCharacter? {
+    // TODO(jwilson): max results = 2
+    val list = list(session)
+    require(list.size <= 1) { "expected at most 1 result but was $list" }
+    return list.firstOrNull()
+  }
+
+  override fun <P : Projection> uniqueResultAs(session: Session, projection: KClass<P>): P? {
```

Comment:
can this just be a default impl on the interface? doesn't seem specific to the entity

Reply:
possibly? I’m reluctant to use default methods but that’s not well motivated

Reply:
probably better to start out without them until we see the same pattern repeating over and over again